### PR TITLE
APPS-2047 NF docker CLI

### PR DESCRIPTION
**Pull request by Gvaihir**
Added `--cache-docker` CLI command to `dx build --nextflow`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ Categories for each release: Added, Changed, Deprecated, Removed, Fixed, Securit
 ### Added
 
 * certifi dxpy dependency
+* docker caching to the CLI commands
```
Comment (YuxinShi0423) [inline CHANGELOG.md:12]: `--cache-docker` for `dx build --nextflow`?

File: src/python/dxpy/nextflow/ImageRef.py [hunk 2]
```diff
@@ -118,13 +118,17 @@ def _cache(self, file_name):
...
+            digest_cmd = "docker images --no-trunc --quiet {}".format(full_image_ref)
+            extracted_digest = subprocess.check_output(digest_cmd, shell=True).decode().strip()
         uploaded_dx_file = upload_local_file(
             filename=file_name,
             project=config["DX_PROJECT_CONTEXT_ID"],
             folder=self._caching_dir,
             name=file_name,
             parents=True,
-            properties={"image_digest": self._digest}
+            properties={"image_digest": self._digest or extracted_digest}
```
Comment (YuxinShi0423) [inline src/python/dxpy/nextflow/ImageRef.py:131]: `self._digest` is already null in this if block, so should be `extraced_digest` only

File: src/python/dxpy/nextflow/ImageRefFactory.py [hunk 3]
```diff
@@ -35,7 +35,7 @@ def get_image(self):
         return image(
             process=self._image_ref["process"],
             digest=self._image_ref["digest"],
-            dx_file_id=self._image_ref.get("dx_file_id", None),
+            dx_file_id=self._image_ref.get("file_id", None),
```
Comment (YuxinShi0423) [inline src/python/dxpy/nextflow/ImageRefFactory.py:38]: is the image ref dict in nextaur also refactored?

File: src/python/dxpy/nextflow/nextflow_builder.py [hunk 4]
```diff
@@ -46,6 +56,7 @@ def parse_extra_args(extra_args):
             dx_input["whats_new"] = extra_args.get("details", {}).get("whatsNew")
         return dx_input
 
+    extra_args = extra_args or {}
```
Comment (YuxinShi0423) [inline src/python/dxpy/nextflow/nextflow_builder.py:59]: might not needed: see https://github.com/dnanexus/dx-toolkit/blob/c3e79d533b8633bc71239172c4b03cf07d5ce33d/src/python/dxpy/scripts/dx_build_app.py#L1211C78-L1211C88

File: src/python/dxpy/nextflow/nextflow_builder.py [hunk 5]
```diff
@@ -58,6 +69,10 @@ def parse_extra_args(extra_args):
         input_hash["github_credentials"] = parse_obj(git_creds, "file")
     if destination:
         build_project_id, build_folder, _ = try_call(resolve_existing_path, destination, expected='folder')
+    if docker_secrets:
+        input_hash["docker_secrets"] = parse_obj(git_creds, "file")
```
Comment (YuxinShi0423) [inline src/python/dxpy/nextflow/nextflow_builder.py:73]: git_creds?

**Review by YuxinShi0423 — CHANGES_REQUESTED** [review]

File: src/python/dxpy/nextflow/ImageRefFactory.py (hunk 3 above)
Comment (Gvaihir) [reply src/python/dxpy/nextflow/ImageRefFactory.py:38]: it was misnamed here. No need

File: src/python/dxpy/nextflow/ImageRef.py (hunk 2 above)
Comment (Gvaihir) [reply src/python/dxpy/nextflow/ImageRef.py:131]: `self._digest` might be an empty string.

File: src/python/dxpy/nextflow/nextflow_builder.py (hunk 4 above)
Comment (Gvaihir) [reply src/python/dxpy/nextflow/nextflow_builder.py:59]: maybe but for whatever reason a default empty dict was given in the earlier version, which is a bad practice. I set default to None which I substitute if needed.

File: CHANGELOG.md (hunk 1 above)
Comment (Gvaihir) [reply CHANGELOG.md]: sure, in [403e1ba](https://github.com/dnanexus/dx-toolkit/pull/1280/commits/403e1ba3f83672c24cbb1c97901f4aae7016a7b6)

File: src/python/dxpy/nextflow/nextflow_builder.py (hunk 5 above)
Comment (Gvaihir) [reply src/python/dxpy/nextflow/nextflow_builder.py]: thanks, [403e1ba](https://github.com/dnanexus/dx-toolkit/pull/1280/commits/403e1ba3f83672c24cbb1c97901f4aae7016a7b6)

File: src/python/dxpy/nextflow/ImageRef.py (hunk 2 above)
Comment (YuxinShi0423) [reply src/python/dxpy/nextflow/ImageRef.py:131]: in line 121 `extracted_digest = self._digest`
so do we really have to pick `self._digest or extracted_digest` ?

File: src/python/dxpy/nextflow/ImageRef.py [hunk 6]
```diff
@@ -133,7 +137,7 @@ def _reconstruct_image_ref(self):
         Docker image reference has the form of <REPOSITORY_NAME>/<IMAGE_NAME>:<VERSION_TAG> or
         <REPOSITORY_NAME>/<IMAGE_NAME>@<DIGEST>
         """
-        repo_and_image_name = self._join_if_exists("/", [self._repository, self._image_name])
+        repo_and_image_name = self._join_if_exists("", [self._repository, self._image_name])
```
Comment (YuxinShi0423) [inline src/python/dxpy/nextflow/ImageRef.py:140]: why the delimiter is removed here?

File: src/python/dxpy/scripts/dx.py [hunk 7]
```diff
@@ -2834,12 +2834,24 @@ def handle_arg_conflicts(args):
...
+        if args.cache_docker and args.remote:
+            build_parser.error("Nextflow pipeline built with an option to cache the docker images is always built using the Nextflow Pipeline Importer app. This is not compatible with --remote.")
+
         if args.git_credentials and not args.repository:
             build_parser.error("Git credentials can be supplied only when building Nextflow pipeline from a Git repository.")
 
         if args.nextflow and args.mode == "app":
             build_parser.error("Building Nextflow apps is not supported. Build applet instead.")
 
+        if args.cache_docker:
```
Comment (YuxinShi0423) [inline src/python/dxpy/scripts/dx.py:2846]: check `if args.cache_docker and not args.nextflow:` as in line 2831?

**Review by YuxinShi0423 — COMMENTED** [review]
Are we having tests for here or in file-apps?

File: src/python/dxpy/nextflow/ImageRef.py (hunk 2 above)
Comment (Gvaihir) [reply src/python/dxpy/nextflow/ImageRef.py:131]: cleaned up

File: src/python/dxpy/nextflow/ImageRef.py (hunk 6 above)
Comment (Gvaihir) [reply src/python/dxpy/nextflow/ImageRef.py:139]: Nextaur CLI outputs json with [repository](https://github.com/dnanexus/nextaur-app/blob/98a2e583286511f50ec9f35c116eef11a66516fb/plugins/nextaur/src/main/nextflow/cloud/dnanexus/fs/DxPathTokens.java#L33) with a trailing slash

File: src/python/dxpy/scripts/dx.py (hunk 7 above)
Comment (Gvaihir) [reply src/python/dxpy/scripts/dx.py:2846]: thanks!
[888fcd1](https://github.com/dnanexus/dx-toolkit/pull/1280/commits/888fcd10eaa623dd0c6ae9ac1aceb85c2bf17aa1)

**Comment by Gvaihir** [comment]
> Are we having tests for here or in file-apps?

[aeef5a5](https://github.com/dnanexus/dx-toolkit/pull/1280/commits/aeef5a55f132811ca398c8403a5725c785ab7ae6)
But more is covered with integration tests at QE. 
This added test will fail before merge to `master` due to NPI applet depending on the functionality being tested

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,6 +9,7 @@ Categories for each release: Added, Changed, Deprecated, Removed, Fixed, Securit
 ### Added
 
 * Python 3.12 support
+* * `--cache-docker` for `dx build --nextflow`
```
Comment (YuxinShi0423) [inline CHANGELOG.md:12]: single indent

File: src/python/test/test_nextflow.py [hunk 9]
```diff
@@ -276,6 +276,26 @@ def test_bundle_docker_images(self):
...
+            pipeline_name, existing_nf_file_path=self.base_nextflow_docker)
+
+        # Override metadata values
+        applet_id = json.loads(run(
+            "dx build --nextflow '{}' --cache-docker".format(applet_dir)))["id"]
+
+        applet = dxpy.DXApplet(applet_id)
+        desc = applet.describe()
+        dependencies = desc.get("runSpec").get("bundledDepends")
+        docker_dependency = [x for x in dependencies if x["name"] == "bash"]
```
Comment (YuxinShi0423) [inline src/python/test/test_nextflow.py:293]: what about the other metadata? is the `name` the only one we can check on?

**Review by YuxinShi0423 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 8 above)
Comment (Gvaihir) [reply CHANGELOG.md]: [27fd155](https://github.com/dnanexus/dx-toolkit/pull/1280/commits/27fd15562c1edc8ad10140540adbe5adb4da375f)

File: src/python/dxpy/nextflow/nextflow_builder.py [hunk 10]
```diff
@@ -17,7 +17,17 @@
...
+        cache_docker=False,
+        docker_secrets=None,
+        profile="",
+        git_creds=None,
+        brief=False,
+        destination=None,
+        extra_args=None
+):
     """
     :param repository: URL to a Git repository
```
Comment (YuxinShi0423) [inline src/python/dxpy/nextflow/nextflow_builder.py:32]: add docstring for new fields?

File: src/python/dxpy/nextflow/nextflow_utils.py [hunk 11]
```diff
@@ -85,11 +94,14 @@ def write_dxapp(folder, content):
...
+def get_regional_options(region, resources_dir, cache_docker):
     nextaur_asset, nextflow_asset = get_nextflow_assets(region)
     regional_instance_type = get_instance_type(region)
-    image_refs = None
-    image_bundled = None
+    if cache_docker:
+        image_refs = run_nextaur_collect(resources_dir)
+        image_bundled = bundle_docker_images(image_refs)
+    else:
+        image_bundled = {}
```
Comment (YuxinShi0423) [inline src/python/dxpy/nextflow/nextflow_utils.py:104]: should be an empty array

File: src/python/dxpy/scripts/dx.py [hunk 12]
```diff
@@ -5079,6 +5095,15 @@ def positive_integer(value):
                                                         " at https://www.nextflow.io/blog/2021/configure-git-repositories-with-nextflow.html.",
                                                    width_adjustment=-24), dest="git_credentials").completer = DXPathCompleter(classes=['file'])
 
+
+nextflow_options.add_argument('--cache-docker', help=fill("Stores a container image tarball in the currently selected project"
```
Comment (YuxinShi0423) [inline src/python/dxpy/scripts/dx.py:5099]: minor: maybe add this argument with `default=False`?

File: src/python/dxpy/scripts/dx_build_app.py [hunk 13]
```diff
@@ -1072,7 +1083,7 @@ def _build_app(args, extra_args):
 
         return output['id']
 
-    if not args.remote and not args.repository:  # building with NF repository is implicitly remote
+    if not args.remote and not build_nf_with_npi:  # building with NF repository is implicitly remote
```
Comment (YuxinShi0423) [inline src/python/dxpy/scripts/dx_build_app.py:1086]:
```suggestion
    if not args.remote and not build_nf_with_npi:  # building with NF repository or cache_docker is done remotely by npi
```

**Review by YuxinShi0423 — APPROVED** [review]

File: src/python/dxpy/scripts/dx.py (hunk 12 above)
Comment (Gvaihir) [reply src/python/dxpy/scripts/dx.py:5099]: other boolean flags don't have a default. Following the convention

File: src/python/dxpy/nextflow/nextflow_builder.py (hunk 10 above)
Comment (Gvaihir) [reply src/python/dxpy/nextflow/nextflow_builder.py:32]: [8fa5344](https://github.com/dnanexus/dx-toolkit/pull/1280/commits/8fa5344430a72b77a305f2d2f72f49b2ea7859e4)

File: src/python/dxpy/nextflow/nextflow_utils.py (hunk 11 above)
Comment (Gvaihir) [reply src/python/dxpy/nextflow/nextflow_utils.py]: [8fa5344](https://github.com/dnanexus/dx-toolkit/pull/1280/commits/8fa5344430a72b77a305f2d2f72f49b2ea7859e4)

File: src/python/dxpy/scripts/dx_build_app.py (hunk 13 above)
Comment (Gvaihir) [reply src/python/dxpy/scripts/dx_build_app.py]: [8fa5344](https://github.com/dnanexus/dx-toolkit/pull/1280/commits/8fa5344430a72b77a305f2d2f72f49b2ea7859e4)

**Review by YuxinShi0423 — APPROVED** [review]

**Pull request merged**